src/envs/pddl_env.py

```diff
@@ -0,0 +1,442 @@
+            for a in pyperplan_problem.goal
+        }
+        # Finalize the task.
+        task = Task(init, goal)
+        return task
+
+    def _strips_operator_to_parameterized_option(
+            self, op: STRIPSOperator) -> ParameterizedOption:
+        name = op.name
+        types = [p.type for p in op.parameters]
+        op_idx = self._ordered_strips_operators.index(op)
+        act_dims = self.action_space.shape[0]
+
+        def policy(s: State, m: Dict, o: Sequence[Object], p: Array) -> Action:
+            del m, p  # unused
+            ordered_objs = list(s)
+            obj_idxs = [ordered_objs.index(obj) for obj in o]
+            act_arr = np.zeros(act_dims, dtype=np.float32)
+            act_arr[0] = op_idx
+            act_arr[1:len(obj_idxs) + 1] = obj_idxs
+            return Action(act_arr)
+
+        def initiable(s: State, m: Dict, o: Sequence[Object],
+                      p: Array) -> bool:
+            del m, p  # unused
+            ground_atoms = _state_to_ground_atoms(s)
+            ground_op = op.ground(tuple(o))
+            return ground_op.preconditions.issubset(ground_atoms)
+
+        return utils.SingletonParameterizedOption(name,
+                                                  policy,
+                                                  types,
+                                                  initiable=initiable)
+
+
+class _FixedTasksPDDLEnv(_PDDLEnv):
+    """An env where tasks are induced by static PDDL problem files."""
+
+    @property
+    @abc.abstractmethod
+    def _pddl_problem_asset_dir(self) -> str:
+        """The name of the problem file directory in assets/pddl."""
+        raise NotImplementedError("Override me!")
+
+    @property
+    @abc.abstractmethod
+    def _train_problem_indices(self) -> List[int]:
+        """The problem file names should be task{idx}.pddl."""
+        raise NotImplementedError("Override me!")
+
+    @property
+    @abc.abstractmethod
+    def _test_problem_indices(self) -> List[int]:
+        """The problem file names should be task{idx}.pddl."""
+        raise NotImplementedError("Override me!")
+
+    @property
+    def _pddl_train_problem_generator(self) -> PDDLProblemGenerator:
+        assert len(self._train_problem_indices) >= CFG.num_train_tasks
+        return _file_problem_generator(self._pddl_problem_asset_dir,
+                                       self._train_problem_indices)
+
+    @property
+    def _pddl_test_problem_generator(self) -> PDDLProblemGenerator:
+        assert len(self._test_problem_indices) >= CFG.num_test_tasks
+        return _file_problem_generator(self._pddl_problem_asset_dir,
+                                       self._test_problem_indices)
+
+
+###############################################################################
+#                              Concrete Envs                                  #
+###############################################################################
+
+
+class _BlocksPDDLEnv(_PDDLEnv):
```

Comment:
isn't this abstract too?

Reply:
yeah. I changed "Concrete" to "Specific" in the heading. I still think it's good to group all the blocks stuff together